synapse/handlers/_base.py

```diff
@@ -93,20 +101,34 @@ def allowed(event, user_id, is_peeking):
             else:
                 membership = None
 
+            # if the user was a member of the room at the time of the event,
+            # they can see it.
             if membership == Membership.JOIN:
                 return True
 
             if event.type == EventTypes.RoomHistoryVisibility:
-                return not is_peeking
+                # XXX why are m.room.history_visibility events special?
```

Comment:
I've talked with Erik, and we believe the reason for this is that it allows clients to actually see that history visibility changes happened, partially because they can't see the event which opens up history visibility itself.

I contend that this is bad, partially because it leaks membership information about otherwise restricted rooms.

I'm fairly convinced that what we want to do is either:

1) Stop special casing it completely. This means that they necessarily won't see a room become shared.
2) Special case it by allowing it to be returned if the state *before or after* the event would allow you to see it, rather than just before as we currently do.